packages/core/src/actors.ts

```diff
@@ -274,48 +158,46 @@ export function createDeferredBehavior<TEvent extends EventObject>(
 
       return undefined;
     },
+    subscribe: (observer) => {
+      observers.add(observer);
+
+      return {
+        unsubscribe: () => {
+          observers.delete(observer);
+        }
+      };
+    },
     initialState: undefined
   };
 
   return behavior;
 }
 
-export function createPromiseBehavior<T, TEvent extends EventObject>(
+export function fromPromise<T, TEvent extends EventObject>(
   lazyPromise: Lazy<PromiseLike<T>>
 ): Behavior<{ type: string }, T | undefined> {
-  const parent = CapturedState.current?.actorRef;
   let canceled = false;
   const observers: Set<Observer<T>> = new Set();
 
   const behavior: Behavior<TEvent, T | undefined> = {
-    transition: (_, event, actorContext) => {
+    transition: (_, event) => {
       switch (event.type) {
         case startSignalType:
           const resolvedPromise = Promise.resolve(lazyPromise());
 
           resolvedPromise.then(
             (response) => {
               if (!canceled) {
-                parent?.send(
-                  toSCXMLEvent(doneInvoke(actorContext.name, response) as any, {
-                    origin: actorContext.self
-                  })
-                );
-
                 observers.forEach((observer) => {
                   observer.next?.(response);
-                  observer.complete?.();
+                  if (observers.has(observer)) {
+                    observer.complete?.();
+                  }
```

Comment:
Somewhat weirdly this will both emit a snapshot and complete immediately - which is not incorrect, just maybe slightly redundant. It's more an observation on my part and not a strong concern about this.